includes/admin/extensions/class-extensions-api.php

```diff
@@ -0,0 +1,169 @@
+<?php
+
+namespace EDD\Admin\Extensions;
+
+class ExtensionsAPI {
+
+	/**
+	 * Gets the product data from the EDD Products API.
+	 *
+	 * @since 2.11.x
+	 * @param array $body    The body for the API request.
+	 * @param int   $item_id The product ID, if querying a single product.
+	 * @return array|false
+	 */
+	public function get_product_data( $body = array(), $item_id = false ) {
+		if ( empty( $body ) ) {
+			if ( empty( $item_id ) ) {
+				return false;
+			}
+			$body = $this->get_api_body( $item_id );
+		}
+		$key = $this->array_key_first( $body );
+		// The option name is created from the first key/value pair of the API "body".
+		$option_name = "edd_extension_{$key}_{$body[ $key ]}_data";
+		$option      = $this->get_stored_extension_data( $option_name );
```

Comment:
Not urgent, so this could wait for a future release instead. But I think we can rework this cache a little to be more error-friendly. Right now you could run into this scenario:

- You get product data. It gets cached.
- Cache expires, so it gets deleted in `get_stored_extension_data()`.
- You do a new API request.
- It fails.
- You get no data for at least an hour.

Here's a way we could improve that:

- You get product data. It gets cached.
- In `get_product_data()` you note that the cache has expired, so you need to do an API request, but you don't actually delete the cache yet.
- You do a new API request.
- It fails.
- Because you didn't delete the cache, you can just extend the timeout.

This way we reduce the chance of someone seeing no product information. Instead we just keep serving them the stale cache until the API request is successful again.

Reply:
I agree that sounds like a better way to handle it--I had thought of that and would prefer to show old data rather than nothing.

Reply:
Reworked the data request in aa99898 to return stale data if there is an API error, and only delete the option if it is stale and new data is retrieved.